modality-probe-debug-collector/src/lib.rs

```diff
@@ -0,0 +1,1139 @@
+use chrono::Utc;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::error::Error;
+use std::fs::File;
+use std::net::SocketAddrV4;
+use std::path::PathBuf;
+use std::rc::Rc;
+use std::sync::mpsc::Receiver;
+use std::thread::sleep;
+use std::time::Duration;
+
+use probe_rs::{MemoryInterface, Session};
+
+use modality_probe::compact_log::CompactLogItem;
+use modality_probe::{EventId, LogicalClock, ProbeId};
+use modality_probe_udp_collector::add_log_report_to_entries;
+use race_buffer::reader::{RaceBufferReader, Snapper};
+use race_buffer::Entry;
+use util::alloc_log_report::LogReport;
+use util::model::{LogEntry, SegmentId, SessionId};
+
+type Result<T> = std::result::Result<T, Box<dyn Error>>;
+
+// NOTE: These may be changed once RaceBuffer is implemented into ekt
+const PROBE_ID_OFFSET: u32 = 0x0;
+const BUF_ADDR_OFFSET: u32 = 0x4;
+const BUF_CAP_OFFSET: u32 = 0x8;
+const WCURS_OFFSET: u32 = 0xc;
+
+/// Configuration for running collector
+#[derive(Debug, PartialEq)]
+pub struct Config {
+    pub session_id: SessionId,
+    pub big_endian: bool,
+    pub attach_target: Option<String>,
+    pub gdb_addr: Option<SocketAddrV4>,
+    pub interval: Duration,
+    pub output_path: PathBuf,
+    pub probe_addrs: Vec<ProbeAddr>,
+}
+
+/// Struct representing a probe address, either the address of the probe itself or of
+/// a pointer to the probe
+#[derive(Debug, PartialEq)]
+pub enum ProbeAddr {
+    Addr(u32),
+    PtrAddr(u32),
+}
+
+/// Trait used to specify backend used to access device memory
+pub trait MemoryReader {
+    fn read32(&mut self, addr: u32) -> Result<u32>;
+}
+
+/// MemoryReader that uses probe-rs to access device memory
+struct ProbeRsReader(Session);
+
+impl MemoryReader for ProbeRsReader {
+    fn read32(&mut self, addr: u32) -> Result<u32> {
+        let mut core = self.0.core(0)?;
+        Ok(core.read_word_32(addr)?)
+    }
+}
+
+/// Struct used to take snapshots of RaceBuffer on device
+struct MemorySnapper {
+    /// Reader used to read device memory
+    mem_reader: Rc<RefCell<dyn MemoryReader>>,
+    /// Address of RaceBuffer backing storage
+    storage_addr: u32,
+    /// Address of RaceBuffer write cursor
+    wcurs_addr: u32,
+}
+
+impl Snapper<CompactLogItem> for MemorySnapper {
```

Comment:
Because this trait has zero doc-comments or discussion in the design doc, it is tricky to determine its intent and whether the implementation at hand is doing the right thing.

Reply:
I will add doc-comments to the trait